components/gantry/oneaxis/setup.go

```diff
@@ -0,0 +1,390 @@
+func newLimitSwitchGantry(
+	ctx context.Context,
+	cfg config.Component,
+	logger golog.Logger,
+	oneAxis *oneAxis,
+) (gantry.LocalGantry, error) {
+	conf, ok := cfg.ConvertedAttributes.(*AttrConfig)
+	if !ok {
+		return nil, rdkutils.NewUnexpectedTypeError(conf, cfg.ConvertedAttributes)
+	}
+
+	if len(conf.LimitSwitchPins) > 2 {
+		return nil, errBadNumLimitSwitches(cfg.Name, len(conf.LimitSwitchPins))
+	}
+
+	g := &limitSwitchGantry{
+		ctx:             ctx,
+		limitSwitchPins: conf.LimitSwitchPins,
+		limitHigh:       *conf.LimitPinEnabled,
+		oAx:             oneAxis,
+		logger:          logger,
+		limitsPolled:    false,
+	}
+
+	if len(g.limitSwitchPins) == 1 && (g.oAx.mmPerRevolution == 0 || g.oAx.lengthMm == 0) {
+		return nil, errDimensionsNotFound(g.oAx.name, g.oAx.mmPerRevolution, g.oAx.mmPerRevolution)
+	}
+
+	g.limitHitChecker()
+
+	return g, nil
+}
+
+func (g *limitSwitchGantry) limitHitChecker() {
+	// lock the mutex associated with this monitoring function
+	g.limitHitMu.Lock()
+	// check if the state is false
+	if !g.limitsPolled {
+		return
+	}
+	// we know we are monitoring the limit switches in this gantry,
+	// return if the limitHits are already being checked
+	g.limitsPolled = true
+	g.limitHitMu.Unlock()
+
+	var errs error
+	for {
+		select {
+		case <-g.ctx.Done():
+			return
+		default:
+		}
+		for idx := range g.limitSwitchPins {
+			hit, err := g.limitHit(g.ctx, idx)
+			if hit || err != nil {
+				g.limitHitMu.Lock()
+				g.limitsPolled = false
+				g.limitHitMu.Unlock()
+				errs := multierr.Combine(errs, err)
+				// motor operation manager cancels running
+				err = g.oAx.motor.Stop(g.ctx, nil)
+				errs = multierr.Combine(errs, err)
+				if errs != nil {
+					g.logger.Error(errs)
+					return
+				}
+			}
+		}
+	}
+}
+
+func (g *limitSwitchGantry) limitHit(ctx context.Context, offset int) (bool, error) {
+	if offset < 0 || offset > 1 {
+		return true, errBadNumLimitSwitches(g.oAx.name, offset)
+	}
+	pin, err := g.oAx.board.GPIOPinByName(g.limitSwitchPins[offset])
+	if err != nil {
+		return false, err
+	}
+	high, err := pin.Get(ctx, nil)
+
+	return high == g.limitHigh, err
+}
+
+func (g *limitSwitchGantry) testLimit(ctx context.Context, limInt int) (float64, error) {
+	defer utils.UncheckedErrorFunc(func() error {
+		return g.oAx.motor.Stop(ctx, nil)
+	})
+
+	d := -1.0 // go backwards to hit first limit switch pin
+	if limInt > 0 {
+		d *= -1 // go forwards to hit second limit switch pin
+	}
+
+	err := g.oAx.motor.GoFor(ctx, d*g.oAx.rpm, 0, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	start := time.Now()
+	for {
+		hit, err := g.limitHit(ctx, limInt)
+		if err != nil {
+			return 0, err
+		}
+		if hit {
+			err = g.oAx.motor.Stop(ctx, nil)
+			if err != nil {
+				return 0, err
+			}
+			break
+		}
+
+		elapsed := start.Sub(start)
+		if elapsed > (time.Second * 15) {
+			return 0, errors.New("gantry timed out testing limit")
+		}
+
+		if !utils.SelectContextOrWait(ctx, time.Millisecond*10) {
```

Comment:
This is basically a busy loop with a small sleep in it to make it not-quite-so-busy. It feels like what we really want is for the limit switch to be a digital interrupt instead of a GPIO pin, so that we can replace the for loop with a select statement waiting on either the context to be canceled, the timeout to occur, or the limit switch to get hit:
```
timeout := time.NewTicker(15 * Time.Second)
defer timeout.Stop()
g.oAx.motor.GoFor(...)
defer g.oAx.motor.Stop(...)

select {
    case <-g.ctx.Done():
        return 0, ctx.Err()
    case <-timeout.C:
        return 0, errors.New("gantry timed out")
    case <-g.limitSwitchHit:
        return g.oAx.motor.Position(...)
}
```
but Gautham's recent work with the ultrasonic sensor makes me question whether digital interrupts have decent performance at all, so maybe this isn't a great approach.

Reply:
No, I take that last part back: Gautham's work required microsecond responses, whereas with this motor it's good enough if you're off by a whole millisecond. I think digital interrupts are likely the way to go when implementing limit switches.

Reply:
Upon closer examination, we found the digital interrupt latency to not be an issue unless you wanted the specific time that a tick was received (and even this problem is being resolved). It shouldn't be an issue here should you use a digital interrupt